File: 543-diameter-of-binary-tree/diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int soln(TreeNode root,int arr[]){
        if(root==null) return 0;
        int lh = soln(root.left,arr);
        int rh = soln(root.right,arr);
        arr[0] = Math.max(arr[0],lh+rh);
        return Math.max(lh,rh)+1;
    }
    public int diameterOfBinaryTree(TreeNode root) {
        int arr[] = new int[1];
        soln(root,arr);
        return arr[0];
    }
}

File: 543-diameter-of-binary-tree/diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
int soln(TreeNode* root,int &ans){
    if(!root) return 0;
    int lh = soln(root->left,ans);
    int rh = soln(root->right,ans);
    ans = max(ans,lh+rh+1);
    return max(lh,rh)+1;
}
class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        int ans = 0;
        soln(root,ans);
        return ans-1;
    }
};